### PR TITLE
Fork aware getBlock by height

### PR DESCRIPTION
**Pull request by m4ksio**
this RP upgrades the get block by height functionality to accepts a header and be fork aware.  This makes the fvm get block calls deterministic.

### CODE REVIEW

File: fvm/blocks.go [hunk 1]
```diff
@@ -1,9 +1,68 @@
...
+	}
+
+	id := header.ParentID
+
+	// travel chain back
+	for {
+		// recent block should be in cache so this is supposed to be fast
+		parent, err := b.storage.ByBlockID(id)
+		if err != nil {
+			return nil, fmt.Errorf("cannot retireve block parent: %w", err)
```
Comment (jordanschalm) [inline fvm/blocks.go:48]:
```suggestion
			return nil, fmt.Errorf("cannot retrieve block parent: %w", err)
```

File: fvm/blocks.go [hunk 2]
```diff
@@ -1,9 +1,68 @@
...
+
+		_, err = b.storage.ByHeight(parent.Height)
+		// if height isn't finalized, move to parent
+		if err != nil && errors.Is(err, storage.ErrNotFound) {
+			id = parent.ParentID
+			continue
+		}
+		// any other error bubbles up
+		if err != nil {
+			return nil, fmt.Errorf("cannot retireve block parent: %w", err)
```
Comment (jordanschalm) [inline fvm/blocks.go:62]:
```suggestion
			return nil, fmt.Errorf("cannot retrieve block parent: %w", err)
```

File: fvm/blocks.go [hunk 3]
```diff
@@ -1,9 +1,68 @@
...
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
 
 type Blocks interface {
-	// ByHeight returns the block at the given height. It is only available
-	// for finalized blocks.
-	ByHeight(height uint64) (*flow.Block, error)
+	// ByHeight returns the block at the given height in chain engine at header,
+	// or finalized if header is nil
```
Comment (jordanschalm) [inline fvm/blocks.go:12]:
```suggestion
	// ByHeight returns the block at the given height in the chain ending in `header` (or finalized
	// if `header` is nil). This enables querying un-finalized blocks by height with respect to the
	// chain defined by the block we are executing.
```

Could add some more about why we need a reference block here

File: fvm/blocks_test.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+		header1 := unittest.BlockHeaderWithParentFixture(&header0)
+		header1.Height = 1
+		header2 := unittest.BlockHeaderWithParentFixture(&header1)
+		header2.Height = 2
+		header3 := unittest.BlockHeaderWithParentFixture(&header2)
+		header3.Height = 3
+		header4 := unittest.BlockHeaderWithParentFixture(&header3)
+		header4.Height = 4
+		header5 := unittest.BlockHeaderWithParentFixture(&header4)
+		header5.Height = 5
```
Comment (jordanschalm) [inline fvm/blocks_test.go:56]:
```suggestion
		header1 := unittest.BlockHeaderWithParentFixture(&header0)
		header2 := unittest.BlockHeaderWithParentFixture(&header1)
		header3 := unittest.BlockHeaderWithParentFixture(&header2)
		header4 := unittest.BlockHeaderWithParentFixture(&header3)
		header5 := unittest.BlockHeaderWithParentFixture(&header4)
```

Should be able to remove these, as it is handled by the fixture. https://github.com/onflow/flow-go/blob/e726e24465c8aa1965f12cdb95cbdb9ea3f01632/utils/unittest/fixtures.go#L245

File: fvm/blocks_test.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		require.NoError(t, err)
+		require.Equal(t, header0, *heightFrom)
+	}))
+
+	t.Run("skips heights once it get to finalized chain", doTest(t, func(t *testing.T, headers *storageMock.Headers, blockFinder Blocks, header flow.Header) {
+
+		header0 := unittest.BlockHeaderFixture()
+		header0.Height = 0
+		header1 := unittest.BlockHeaderWithParentFixture(&header0)
+		header1.Height = 1
```
Comment (jordanschalm) [inline fvm/blocks_test.go:79]: Likewise here, the fixture will set heights consistently

**Review by jordanschalm — APPROVED** [review]
LGTM 💯

File: fvm/blocks_test.go (hunk 5 above)
Comment (ramtinms) [reply fvm/blocks_test.go]: Applied at 6202356

**Review by ramtinms — APPROVED** [review]

**Pull request merged**